src/main/java/org/stellar/sdk/requests/SSEStream.java

```diff
@@ -0,0 +1,180 @@
+package org.stellar.sdk.requests;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.internal.sse.RealEventSource;
+import okhttp3.sse.EventSource;
+import okhttp3.sse.EventSourceListener;
+import org.stellar.sdk.responses.GsonSingleton;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.net.SocketException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+public class SSEStream<T extends org.stellar.sdk.responses.Response> implements Closeable {
+  private final OkHttpClient okHttpClient;
+  private final RequestBuilder requestBuilder;
+  private final Class<T> responseClass;
+  private final EventListener<T> listener;
+
+  private final AtomicBoolean isStopped = new AtomicBoolean(false);
+  private final AtomicBoolean serverSideClosed = new AtomicBoolean(true); // make sure we start correctly
+  private final AtomicReference<String> pagingToken = new AtomicReference<String>("0");
+  private final AtomicReference<String> lastEventId = new AtomicReference<String>(null);
+  private ExecutorService executorService;
+  private EventSource eventSource = null;
+
+  private SSEStream(final OkHttpClient okHttpClient, final RequestBuilder requestBuilder, final Class<T> responseClass, final EventListener<T> listener) {
+    this.okHttpClient = okHttpClient;
+    this.requestBuilder = requestBuilder;
+    this.responseClass = responseClass;
+    this.listener = listener;
+
+    executorService = Executors.newSingleThreadExecutor();
+    requestBuilder.buildUri(); // call this once to add the segments
+  }
+
+  private void start() {
+    if (isStopped.get()) {
+      throw new IllegalStateException("Already stopped");
+    }
+    executorService.submit(new Runnable() {
+      @Override
+      public void run() {
+        while (!isStopped.get()) {
+          try {
+            Thread.sleep(200);
+            if (serverSideClosed.get() && !isStopped.get()) {
+              serverSideClosed.set(false); // don't restart until true again
+              restart();
```

Comment:
I think that the following sequence is possible:
1. `serverSideClosed.set(false);`
2. Someone calls `SSEStream.close()`
3. `restart();`

It will leave the stream running in the background. I think synchronization will be much easier to read and understand in this class with `synchronized` block. If I'm not missing anything you need to ensure that:
* Closing the stream,
* Restarting the stream,
* Setting `lastEventId`

are atomic. But to be honest, I'll be totally OK if this class is not thread safe (in outside use) if we add a proper docs. In most use-cases there's a single thread that controls the stream (can close the stream).

Reply:
I've added a lock and double check the stopped state now; shouldn't hurt anyone. Together with the atomic state changes, that should take care of a lot of stuff that will indeed probably never happen. But I've done this stuff long enough to err on the safe side here.

Kind of nasty that we need threads here at all. Probably a bad idea on Android. In kotlin or similar languages, you'd probably use a sequence or co-routines to avoid threading. 

A pure Kotlin client might actually be a good idea since it could address native android/ios, web (via wasm), android, and Java. Just a thought.